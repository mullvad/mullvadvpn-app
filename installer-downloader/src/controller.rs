//! This module implements the actual logic performed by different UI components.

use crate::delegate::{AppDelegate, AppDelegateQueue};
use crate::resource;
use crate::ui_downloader::{UiAppDownloader, UiProgressUpdater};

use mullvad_update::api::Version;
use mullvad_update::{
    api::{self, VersionInfoProvider},
    app::{self, AppDownloaderFactory, AppDownloaderParameters},
};

use std::future::Future;
use tokio::sync::{mpsc, oneshot};

/// Actions handled by an async worker task in [handle_action_messages].
enum TaskMessage {
    SetVersionInfo(api::VersionInfo),
    BeginDownload,
    Cancel,
}

/// See the [module-level docs](self).
pub struct AppController {}

/// Public entry function for registering a [AppDelegate].
pub fn initialize_controller<T: AppDelegate + 'static>(delegate: &mut T) {
    use mullvad_update::api::LatestVersionInfoProvider;
    use mullvad_update::app::HttpAppDownloader;

    // App downloader (factory) to use
    type DownloaderFactory<T> = HttpAppDownloader<UiProgressUpdater<T>, UiProgressUpdater<T>>;
    // Version info provider to use
    type VersionInfoProvider = LatestVersionInfoProvider;

    AppController::initialize::<_, DownloaderFactory<T>, VersionInfoProvider>(delegate)
}

impl AppController {
    /// Initialize [AppController] using the provided delegate.
    ///
    /// Providing the downloader and version info fetcher as type arguments, they're decoupled from
    /// the logic of [AppController], allowing them to be mocked.
    pub fn initialize<Delegate, DownloaderFactory, VersionProvider>(delegate: &mut Delegate)
    where
        Delegate: AppDelegate + 'static,
        VersionProvider: VersionInfoProvider + 'static,
        DownloaderFactory: AppDownloaderFactory<
                SigProgress = UiProgressUpdater<Delegate>,
                AppProgress = UiProgressUpdater<Delegate>,
            > + 'static,
    {
        delegate.hide_download_progress();
        delegate.hide_download_button();
        delegate.hide_cancel_button();

        let (task_tx, task_rx) = mpsc::channel(1);
        tokio::spawn(handle_action_messages::<Delegate, DownloaderFactory>(
            delegate.queue(),
            task_rx,
        ));
        delegate.set_status_text(resource::FETCH_VERSION_DESC);
        tokio::spawn(fetch_app_version_info::<Delegate, VersionProvider>(
            delegate,
            task_tx.clone(),
        ));
        Self::register_user_action_callbacks(delegate, task_tx);
    }

    fn register_user_action_callbacks<T: AppDelegate + 'static>(
        delegate: &mut T,
        task_tx: mpsc::Sender<TaskMessage>,
    ) {
        let tx = task_tx.clone();
        delegate.on_download(move || {
            let _ = tx.try_send(TaskMessage::BeginDownload);
        });
        let tx = task_tx.clone();
        delegate.on_cancel(move || {
            let _ = tx.try_send(TaskMessage::Cancel);
        });
    }
}

/// Background task that fetches app version data.
fn fetch_app_version_info<Delegate, VersionProvider>(
    delegate: &mut Delegate,
    download_tx: mpsc::Sender<TaskMessage>,
) -> impl Future<Output = ()>
where
    Delegate: AppDelegate,
    VersionProvider: VersionInfoProvider,
{
    let queue = delegate.queue();

    async move {
        // TODO: handle errors, retry
        let Ok(version_info) = VersionProvider::get_version_info().await else {
            queue.queue_main(move |self_| {
                self_.set_status_text("Failed to fetch version info");
            });
            return;
        };
        let _ = download_tx.try_send(TaskMessage::SetVersionInfo(version_info));
    }
}

/// Async worker that handles actions such as initiating a download, cancelling it, and updating
/// labels.
async fn handle_action_messages<Delegate, DownloaderFactory>(
    queue: Delegate::Queue,
    mut rx: mpsc::Receiver<TaskMessage>,
) where
    Delegate: AppDelegate + 'static,
    DownloaderFactory: AppDownloaderFactory<
        SigProgress = UiProgressUpdater<Delegate>,
        AppProgress = UiProgressUpdater<Delegate>,
    >,
{
    let mut version_info = None;
    let mut active_download = None;

    while let Some(msg) = rx.recv().await {
        match msg {
            TaskMessage::SetVersionInfo(new_version_info) => {
                let version_label = format_latest_version(&new_version_info.stable);
                queue.queue_main(move |self_| {
                    self_.set_status_text(&version_label);
                    self_.show_download_button();
                });
                version_info = Some(new_version_info);
            }
            TaskMessage::BeginDownload => {
                if active_download.is_some() {
                    continue;
                }
                let Some(version_info) = version_info.clone() else {
                    continue;
                };

                let (tx, rx) = oneshot::channel();
                queue.queue_main(move |self_| {
                    // TODO: Select appropriate URLs
                    let Some(app_url) = version_info.stable.urls.first() else {
                        return;
                    };
                    let Some(signature_url) = version_info.stable.signature_urls.first() else {
                        return;
                    };
                    let app_size = version_info.stable.size;

                    self_.set_status_text("");
                    self_.hide_download_button();
                    self_.show_cancel_button();
                    self_.show_download_progress();

                    let downloader = DownloaderFactory::new_downloader(AppDownloaderParameters {
                        signature_url: signature_url.to_owned(),
                        app_url: app_url.to_owned(),
                        app_size,
                        sig_progress: UiProgressUpdater::new(self_.queue()),
                        app_progress: UiProgressUpdater::new(self_.queue()),
                    });

                    let ui_downloader = UiAppDownloader::new(self_, downloader);
                    let _ = tx.send(tokio::spawn(async move {
                        let _ = app::install_and_upgrade(ui_downloader).await;
                    }));
                });
                active_download = rx.await.ok();
            }
            TaskMessage::Cancel => {
                let Some(active_download) = active_download.take() else {
                    continue;
                };
                active_download.abort();
                let _ = active_download.await;

                let version_label = if let Some(version_info) = &version_info {
                    format_latest_version(&version_info.stable)
                } else {
                    "".to_owned()
                };

                queue.queue_main(move |self_| {
                    self_.set_status_text(&version_label);
                    self_.show_download_button();
                    self_.hide_cancel_button();
                    self_.hide_download_progress();
                    self_.set_download_progress(0);
                });
            }
        }
    }
}

fn format_latest_version(version: &Version) -> String {
    format!("{}: {}", resource::LATEST_VERSION_PREFIX, version.version)
}
