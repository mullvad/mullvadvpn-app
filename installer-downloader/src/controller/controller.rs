//! This module implements the actual logic performed by different UI components.

use super::{AppDelegate, AppDelegateQueue};

use crate::{
    api::{self, LatestVersionInfoProvider, VersionInfoProvider},
    app::{self, AppDownloader, AppDownloaderFactory, AppDownloaderParameters, HttpAppDownloader},
    fetch::ProgressUpdater,
};

use std::future::Future;
use tokio::sync::{mpsc, oneshot};

use super::ui_downloader::{UiAppDownloader, UiProgressUpdater};

/// This trait glues together different components needed to construct an app controller.
/// In particular, this lets us mock the downloader and version provider independently of the
/// UI implementation (some [AppDelegate]).
pub trait AppControllerProvider {
    type Delegate: AppDelegate + 'static;
    type DownloaderFactory: AppDownloaderFactory;
    type VersionInfoProvider: VersionInfoProvider;
}

/// Default implementation of [AppControllerProvider], using an actual HTTP client, and API version
/// fetcher.
pub struct DefaultAppControllerProvider<T> {
    _phantom: std::marker::PhantomData<T>,
}

impl<T> DefaultAppControllerProvider<T> {
    pub fn new() -> Self {
        Self {
            _phantom: std::marker::PhantomData,
        }
    }
}

impl<T: AppDelegate + 'static> AppControllerProvider for DefaultAppControllerProvider<T> {
    type Delegate = T;
    type DownloaderFactory = HttpAppDownloader<UiProgressUpdater<T>, UiProgressUpdater<T>>;
    type VersionInfoProvider = LatestVersionInfoProvider;
}

/// Actions handled by an async worker task in [handle_action_messages].
enum TaskMessage {
    SetVersionInfo(api::VersionInfo),
    BeginDownload,
    Cancel,
}

/// See the [module-level docs](self).
pub struct AppController {}

impl AppController {
    /// Initialize the app controller.
    pub fn initialize<T: AppControllerProvider + 'static>(delegate: &mut T::Delegate) {
        delegate.hide_download_button();
        delegate.hide_cancel_button();

        let (task_tx, task_rx) = mpsc::channel(1);
        tokio::spawn(handle_action_messages::<T>(delegate.queue(), task_rx));
        tokio::spawn(fetch_app_version_info::<T>(delegate, task_tx.clone()));
        Self::register_user_action_callbacks(delegate, task_tx);
    }

    fn register_user_action_callbacks<T: AppDelegate + 'static>(
        delegate: &mut T,
        task_tx: mpsc::Sender<TaskMessage>,
    ) {
        let tx = task_tx.clone();
        delegate.on_download(move |_delegate| {
            let _ = tx.try_send(TaskMessage::BeginDownload);
        });
        let tx = task_tx.clone();
        delegate.on_cancel(move |_delegate| {
            let _ = tx.try_send(TaskMessage::Cancel);
        });
    }
}

/// Background task that fetches app version data.
fn fetch_app_version_info<T: AppControllerProvider>(
    delegate: &mut T::Delegate,
    download_tx: mpsc::Sender<TaskMessage>,
) -> impl Future<Output = ()> {
    delegate.set_status_text("Fetching app version...");
    let queue = delegate.queue();

    async move {
        // TODO: handle errors, retry
        let Ok(version_info) = T::VersionInfoProvider::get_version_info().await else {
            queue.queue_main(move |self_| {
                self_.set_status_text("Failed to fetch version info");
            });
            return;
        };
        let _ = download_tx.try_send(TaskMessage::SetVersionInfo(version_info));
    }
}

/// Async worker that handles actions such as initiating a download, cancelling it, and updating
/// labels.
async fn handle_action_messages<T: AppControllerProvider>(
    queue: <T::Delegate as AppDelegate>::Queue,
    mut rx: mpsc::Receiver<TaskMessage>,
) {
    let mut version_info = None;
    let mut active_download = None;

    while let Some(msg) = rx.recv().await {
        match msg {
            TaskMessage::SetVersionInfo(new_version_info) => {
                let version_label = format!("Latest version: {}", new_version_info.stable.version);
                queue.queue_main(move |self_| {
                    self_.set_status_text(&version_label);
                    self_.show_download_button();
                });
                version_info = Some(new_version_info);
            }
            TaskMessage::BeginDownload => {
                if active_download.is_some() {
                    continue;
                }
                let Some(version_info) = version_info.clone() else {
                    continue;
                };

                let (tx, rx) = oneshot::channel();
                queue.queue_main(move |self_| {
                    // TODO: Select appropriate URLs
                    let Some(app_url) = version_info.stable.urls.first() else {
                        return;
                    };
                    let Some(signature_url) = version_info.stable.signature_urls.first() else {
                        return;
                    };
                    let app_size = version_info.stable.size;

                    self_.set_status_text("");
                    self_.disable_download_button();
                    self_.show_cancel_button();
                    self_.show_download_progress();

                    let new_delegated_downloader = |sig_progress, app_progress| {
                        HttpAppDownloader::new(AppDownloaderParameters {
                            signature_url: signature_url.to_owned(),
                            app_url: app_url.to_owned(),
                            app_size,
                            sig_progress,
                            app_progress,
                        })
                    };

                    let downloader = UiAppDownloader::new(self_, new_delegated_downloader);
                    let _ = tx.send(tokio::spawn(async move {
                        let _ = app::install_and_upgrade(downloader).await;
                    }));
                });
                active_download = rx.await.ok();
            }
            TaskMessage::Cancel => {
                let Some(active_download) = active_download.take() else {
                    continue;
                };
                active_download.abort();
                let _ = active_download.await;

                let version_label = if let Some(version_info) = &version_info {
                    format!("Latest version: {}", version_info.stable.version)
                } else {
                    "".to_owned()
                };

                queue.queue_main(move |self_| {
                    self_.set_status_text(&version_label);
                    self_.enable_download_button();
                    self_.hide_cancel_button();
                    self_.hide_download_progress();
                    self_.set_download_progress(0);
                });
            }
        }
    }
}
