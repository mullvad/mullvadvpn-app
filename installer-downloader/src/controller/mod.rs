//! Framework-agnostic module that hooks up a UI to actions

mod controller;
mod ui_downloader;

pub use ui_downloader::UiProgressUpdater;

use crate::{api::LatestVersionInfoProvider, app::HttpAppDownloader};

/// Trait implementing high-level UI actions
pub trait AppDelegate {
    /// Queue lets us perform actions from other threads
    type Queue: AppDelegateQueue<Self>;

    /// Register click handler for the download button
    fn on_download<F>(&mut self, callback: F)
    where
        F: Fn() + Send + 'static;

    /// Register click handler for the cancel button
    fn on_cancel<F>(&mut self, callback: F)
    where
        F: Fn() + Send + 'static;

    /// Set download status text
    fn set_status_text(&mut self, text: &str);

    /// Show download progress bar
    fn show_download_progress(&mut self);

    /// Hide download progress bar
    fn hide_download_progress(&mut self);

    /// Update download progress bar
    fn set_download_progress(&mut self, complete: u32);

    /// Enable download button
    fn enable_download_button(&mut self);

    /// Disable download button
    fn disable_download_button(&mut self);

    /// Show download button
    fn show_download_button(&mut self);

    /// Hide download button
    fn hide_download_button(&mut self);

    /// Show cancel button
    fn show_cancel_button(&mut self);

    /// Hide cancel button
    fn hide_cancel_button(&mut self);

    /// Create queue for scheduling actions on UI thread
    fn queue(&self) -> Self::Queue;
}

/// Schedules actions on the UI thread from other threads
pub trait AppDelegateQueue<T: ?Sized>: Send {
    fn queue_main<F: FnOnce(&mut T) + 'static + Send>(&self, callback: F);
}

pub use controller::AppController;

/// Public entry function for registering a [AppDelegate].
pub fn initialize_controller<T: AppDelegate + 'static>(delegate: &mut T) {
    // App downloader (factory) to use
    type DownloaderFactory<T> = HttpAppDownloader<UiProgressUpdater<T>, UiProgressUpdater<T>>;
    // Version info provider to use
    type VersionInfoProvider = LatestVersionInfoProvider;

    controller::AppController::initialize::<_, DownloaderFactory<T>, VersionInfoProvider>(delegate)
}
