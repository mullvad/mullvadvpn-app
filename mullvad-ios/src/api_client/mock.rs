use mockito::{Mock, ServerGuard};
use std::ffi::c_void;

#[repr(C)]
pub struct SwiftServerMock {
    server_ptr: *const c_void,
    mock_ptr: *const c_void,
    port: u16,
}

impl SwiftServerMock {
    pub fn new(server: ServerGuard, mock: Mock) -> SwiftServerMock {
        let port = server.socket_address().port();
        let server_ptr = Box::into_raw(Box::new(server)) as *const c_void;
        let mock_ptr = Box::into_raw(Box::new(mock)) as *const c_void;

        SwiftServerMock {
            server_ptr,
            mock_ptr,
            port,
        }
    }
}

/// # Safety
///
/// `method` must be a pointer to a null terminated string representing the http method.
///
/// `path` must be a pointer to a null terminated string representing the url path.
///
/// `response_code` must be a usize representing the http response code.
///
/// `response_body` must be a pointer to a null terminated string representing the body.
///
/// This function is safe.
#[unsafe(no_mangle)]
pub unsafe extern "C" fn mullvad_api_mock_get(
    path: *const u8,
    response_code: usize,
    response_body: *const u8,
) -> SwiftServerMock {
    let path = unsafe { std::ffi::CStr::from_ptr(path.cast()) }
        .to_str()
        .unwrap();
    let response_body = unsafe { std::ffi::CStr::from_ptr(response_body.cast()) }
        .to_str()
        .unwrap();
    let mut server = mockito::Server::new();
    let mock = server
        .mock("GET", path)
        .with_header("content-type", "application/json")
        .with_status(response_code)
        .with_body(response_body)
        .create();
    SwiftServerMock::new(server, mock)
}

/// # Safety
///
/// `path` must be a pointer to a null terminated string representing the url path.
///
/// `response_code` must be a usize representing the http response code.
///
/// `match_body` must be a pointer to a null terminated string representing the body the server expects.
///
/// This function is safe.
#[unsafe(no_mangle)]
pub unsafe extern "C" fn mullvad_api_mock_post(
    path: *const u8,
    response_code: usize,
    match_body: *const u8,
) -> SwiftServerMock {
    let path = unsafe { std::ffi::CStr::from_ptr(path.cast()) }
        .to_str()
        .unwrap();
    let match_body = unsafe { std::ffi::CStr::from_ptr(match_body.cast()) }
        .to_str()
        .unwrap();
    let mut server = mockito::Server::new();
    let mock = server
        .mock("POST", path)
        .with_header("content-type", "application/json")
        .with_status(response_code)
        .match_body(mockito::Matcher::JsonString(match_body.to_string()))
        .create();
    SwiftServerMock::new(server, mock)
}

/// Called by the Swift side to signal that the Rust `SwiftServerMock` can be safely
/// dropped from memory.
///
/// # Safety
///
/// `mock_ptr` must be pointing to a valid instance of `SwiftServerMock`. This function
/// is not safe to call multiple times with the same `SwiftServerMock`.
#[no_mangle]
extern "C" fn mullvad_api_mock_drop(mock_ptr: SwiftServerMock) {
    if !mock_ptr.mock_ptr.is_null() {
        unsafe { drop(Box::from_raw(mock_ptr.mock_ptr as *mut Mock)) };
    }
    if !mock_ptr.server_ptr.is_null() {
        unsafe { drop(Box::from_raw(mock_ptr.server_ptr as *mut ServerGuard)) };
    }
}
