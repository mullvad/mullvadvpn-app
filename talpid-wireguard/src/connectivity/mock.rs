use std::future::Future;
use std::pin::Pin;
use tokio::time::Instant;

use super::check::{CancelToken, ConnState, PingState};
use super::pinger;
use super::Check;

use crate::{Config, Tunnel, TunnelError};
use pinger::Pinger;

// Convenient re-exports
pub use crate::stats::{Stats, StatsMap};

#[derive(Default)]
pub(crate) struct MockPinger {
    on_send_ping: Option<Box<dyn FnMut() + Send + Sync>>,
}

pub(crate) struct MockTunnel {
    on_get_stats: Box<dyn Fn() -> Result<StatsMap, TunnelError> + Send + Sync>,
}

pub fn mock_checker(now: Instant, pinger: Box<dyn Pinger>) -> (Check, CancelToken) {
    let conn_state = ConnState::new(now, Default::default());
    let ping_state = PingState::new_with(pinger);
    Check::mock(conn_state, ping_state)
}

pub fn connected_state(timestamp: Instant) -> ConnState {
    const PEER: [u8; 32] = [0u8; 32];
    let mut stats = StatsMap::new();
    stats.insert(
        PEER,
        Stats {
            tx_bytes: 0,
            rx_bytes: 0,
        },
    );
    ConnState::Connected {
        rx_timestamp: timestamp,
        tx_timestamp: timestamp,
        stats,
    }
}

impl MockTunnel {
    const PEER: [u8; 32] = [0u8; 32];

    pub fn new<F: Fn() -> Result<StatsMap, TunnelError> + Send + Sync + 'static>(f: F) -> Self {
        Self {
            on_get_stats: Box::new(f),
        }
    }

    /// Convert self to the more general [TunnelType].
    pub fn boxed(self) -> Box<dyn Tunnel> {
        Box::new(self)
    }

    pub fn always_incrementing() -> Self {
        let mut map = StatsMap::new();
        map.insert(
            Self::PEER,
            Stats {
                tx_bytes: 0,
                rx_bytes: 0,
            },
        );
        let peers = std::sync::Mutex::new(map);
        Self {
            on_get_stats: Box::new(move || {
                let mut peers = peers.lock().unwrap();
                for traffic in peers.values_mut() {
                    traffic.tx_bytes += 1;
                    traffic.rx_bytes += 1;
                }
                Ok(peers.clone())
            }),
        }
    }

    pub fn never_incrementing() -> Self {
        Self {
            on_get_stats: Box::new(|| {
                let mut map = StatsMap::new();
                map.insert(
                    Self::PEER,
                    Stats {
                        tx_bytes: 0,
                        rx_bytes: 0,
                    },
                );
                Ok(map)
            }),
        }
    }
}

#[async_trait::async_trait]
impl Tunnel for MockTunnel {
    fn get_interface_name(&self) -> String {
        "mock-tunnel".to_string()
    }

    fn stop(self: Box<Self>) -> Result<(), TunnelError> {
        Ok(())
    }

    async fn get_tunnel_stats(&self) -> Result<StatsMap, TunnelError> {
        (self.on_get_stats)()
    }

    fn set_config(
        &mut self,
        _config: Config,
    ) -> Pin<Box<dyn Future<Output = std::result::Result<(), TunnelError>> + Send>> {
        Box::pin(async { Ok(()) })
    }

    #[cfg(daita)]
    fn start_daita(
        &mut self,
        _: talpid_tunnel_config_client::DaitaSettings,
    ) -> std::result::Result<(), TunnelError> {
        Ok(())
    }
}

#[async_trait::async_trait]
impl Pinger for MockPinger {
    async fn send_icmp(&mut self) -> Result<(), pinger::Error> {
        if let Some(callback) = self.on_send_ping.as_mut() {
            (callback)();
        }
        Ok(())
    }
}
