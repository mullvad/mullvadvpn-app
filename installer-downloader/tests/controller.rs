#![cfg(any(target_os = "windows", target_os = "macos"))]

//! Tests for integrations between UI controller and other components
//!
//! The tests rely on `insta` for snapshot testing. If they fail due to snapshot assertions,
//! then most likely the snapshots need to be updated. The most convenient way to review
//! changes to, and update, snapshots is by running `cargo insta review`.

use insta::assert_yaml_snapshot;
use installer_downloader::controller::AppController;
use mock::{
    FakeAppDelegate, FakeAppDownloaderHappyPath, FakeAppDownloaderVerifyFail,
    FakeDirectoryProvider, FakeVersionInfoProvider, FAKE_ENVIRONMENT,
};
use std::{
    sync::{atomic::AtomicBool, Arc},
    time::Duration,
};

mod mock;

/// Test that the flow starts by fetching app version data
#[tokio::test(start_paused = true)]
async fn test_fetch_version() {
    let mut delegate = FakeAppDelegate::default();
    AppController::initialize::<_, FakeAppDownloaderHappyPath, _, FakeDirectoryProvider<true>>(
        &mut delegate,
        FakeVersionInfoProvider::default(),
        FAKE_ENVIRONMENT,
    );

    // The app should start out by fetching the current app version
    assert_yaml_snapshot!(delegate.state);

    tokio::time::sleep(Duration::from_secs(1)).await;

    // Run UI updates to display the fetched version
    let queue = delegate.queue.clone();
    queue.run_callbacks(&mut delegate);

    tokio::time::sleep(Duration::from_secs(1)).await;

    let queue = delegate.queue.clone();
    queue.run_callbacks(&mut delegate);

    // The download button and current version should be displayed
    assert_yaml_snapshot!(delegate.state);
}

/// Test that the on_download callback gets registered and, when invoked,
/// properly updates the UI.
#[tokio::test(start_paused = true)]
async fn test_download() {
    let mut delegate = FakeAppDelegate::default();
    AppController::initialize::<_, FakeAppDownloaderHappyPath, _, FakeDirectoryProvider<true>>(
        &mut delegate,
        FakeVersionInfoProvider::default(),
        FAKE_ENVIRONMENT,
    );

    // Wait for the version info
    tokio::time::sleep(Duration::from_secs(1)).await;

    let queue = delegate.queue.clone();
    queue.run_callbacks(&mut delegate);

    // The download button should be available
    assert_yaml_snapshot!(delegate.state);

    // Initiate download
    let cb = delegate
        .download_callback
        .take()
        .expect("no download callback registered");
    cb();

    tokio::time::sleep(Duration::from_secs(1)).await;

    // Run queued actions
    let queue = delegate.queue.clone();
    queue.run_callbacks(&mut delegate);

    // We should see download progress, and cancellation
    assert_yaml_snapshot!(delegate.state);

    // Wait for download
    tokio::time::sleep(Duration::from_secs(1)).await;

    let queue = delegate.queue.clone();
    queue.run_callbacks(&mut delegate);

    // Everything including verification should have succeeded
    // Downloader should have quit
    assert_yaml_snapshot!(delegate.state);
}

/// Test that the flow of retrying the version fetch after a failure
#[tokio::test(start_paused = true)]
async fn test_failed_fetch_version() {
    let mut delegate = FakeAppDelegate::default();
    let fail_fetching = Arc::new(AtomicBool::new(true));
    AppController::initialize::<_, FakeAppDownloaderHappyPath, _, FakeDirectoryProvider<true>>(
        &mut delegate,
        FakeVersionInfoProvider {
            fail_fetching: fail_fetching.clone(),
        },
        FAKE_ENVIRONMENT,
    );

    tokio::time::sleep(Duration::from_secs(1)).await;

    // Run UI updates to display the fetched version
    let queue = delegate.queue.clone();
    queue.run_callbacks(&mut delegate);

    tokio::time::sleep(Duration::from_secs(1)).await;
    queue.run_callbacks(&mut delegate);

    // The fetch version failure screen with a retry and cancel button should be displayed
    assert_yaml_snapshot!(delegate.state);

    fail_fetching.store(false, std::sync::atomic::Ordering::SeqCst);

    // Retry fetching the version
    let cb = delegate
        .error_retry_callback
        .take()
        .expect("no retry callback registered");
    cb();

    tokio::time::sleep(Duration::from_secs(1)).await;

    // Run UI updates to display the fetched version
    let queue = delegate.queue.clone();
    queue.run_callbacks(&mut delegate);
    tokio::time::sleep(Duration::from_secs(1)).await;
    queue.run_callbacks(&mut delegate);

    // The download button and current version should be displayed
    assert_yaml_snapshot!(delegate.state);
}

/// Test that the install aborts if verification fails
#[tokio::test(start_paused = true)]
async fn test_failed_verification() {
    let mut delegate = FakeAppDelegate::default();
    AppController::initialize::<_, FakeAppDownloaderVerifyFail, _, FakeDirectoryProvider<true>>(
        &mut delegate,
        FakeVersionInfoProvider::default(),
        FAKE_ENVIRONMENT,
    );

    // Wait for the version info
    tokio::time::sleep(Duration::from_secs(1)).await;

    let queue = delegate.queue.clone();
    queue.run_callbacks(&mut delegate);

    // Initiate download
    let cb = delegate
        .download_callback
        .take()
        .expect("no download callback registered");
    cb();

    tokio::time::sleep(Duration::from_secs(1)).await;

    // Wait for queued actions to complete
    let queue = delegate.queue.clone();
    queue.run_callbacks(&mut delegate);

    tokio::time::sleep(Duration::from_secs(1)).await;

    let queue = delegate.queue.clone();
    queue.run_callbacks(&mut delegate);

    // Verification failed
    assert_yaml_snapshot!(delegate.state);
}

/// Test failing to create the download directory
#[tokio::test(start_paused = true)]
async fn test_failed_directory_creation() {
    let mut delegate = FakeAppDelegate::default();
    AppController::initialize::<_, FakeAppDownloaderHappyPath, _, FakeDirectoryProvider<false>>(
        &mut delegate,
        FakeVersionInfoProvider::default(),
        FAKE_ENVIRONMENT,
    );

    // Wait for the version info
    tokio::time::sleep(Duration::from_secs(1)).await;

    let queue = delegate.queue.clone();
    queue.run_callbacks(&mut delegate);

    // Initiate download
    let cb = delegate
        .download_callback
        .take()
        .expect("no download callback registered");
    cb();

    tokio::time::sleep(Duration::from_secs(1)).await;

    // Wait for queued actions to complete
    let queue = delegate.queue.clone();
    queue.run_callbacks(&mut delegate);

    tokio::time::sleep(Duration::from_secs(1)).await;

    let queue = delegate.queue.clone();
    queue.run_callbacks(&mut delegate);

    // "Download failed"
    assert_yaml_snapshot!(delegate.state);
}
