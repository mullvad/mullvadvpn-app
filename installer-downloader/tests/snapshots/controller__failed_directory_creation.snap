---
source: installer-downloader/tests/controller.rs
expression: delegate.state
---
status_text: ""
download_text: ""
download_button_visible: false
cancel_button_visible: false
cancel_button_enabled: false
download_button_enabled: true
download_progress: 0
download_progress_visible: false
beta_text_visible: false
stable_text_visible: false
error_message_visible: true
error_message:
  status_text: "Download failed, please check your internet connection or if you have enough space on your hard drive and try downloading again."
  cancel_button_text: Cancel
  retry_button_text: Try again
quit: false
call_log:
  - hide_download_progress
  - show_download_button
  - disable_download_button
  - hide_cancel_button
  - hide_beta_text
  - hide_stable_text
  - on_download
  - on_cancel
  - on_beta_link
  - on_stable_link
  - show_download_button
  - "set_status_text: Loading version details..."
  - hide_error_message
  - "set_status_text: Version: 2025.1"
  - enable_download_button
  - hide_error_message
  - on_error_message_retry
  - on_error_message_cancel
  - clear_status_text
  - hide_download_button
  - hide_beta_text
  - hide_stable_text
  - "show_error_message: Download failed, please check your internet connection or if you have enough space on your hard drive and try downloading again.. retry: Try again. cancel: Cancel"
